File: src/main/java/com/teamwizardry/wizardry/api/spell/PatternShape.java
package com.teamwizardry.wizardry.api.spell;

import net.minecraft.world.World;

public abstract class PatternShape extends Pattern
{
    @Override
    public void affectEntity(World world, Interactor entity, Instance instance)
    {
        if (instance instanceof ShapeInstance)
        {
            ShapeInstance shape = (ShapeInstance) instance;
            shape.effects.forEach(effect -> effect.run(world, entity));
            shape.nextShape.run(world, entity);
        }
    }
    
    @Override
    public void affectBlock(World world, Interactor block, Instance instance)
    {
        if (instance instanceof ShapeInstance)
        {
            ShapeInstance shape = (ShapeInstance) instance;
            shape.effects.forEach(effect -> effect.run(world, block));
            shape.nextShape.run(world, block);
        }
    }
}

File: src/main/java/com/teamwizardry/wizardry/common/spell/component/PatternShape.kt
package com.teamwizardry.wizardry.common.spell.component

import net.minecraft.world.World

/**
 * Shape effects have affectEntity and affectBlock to run their linked effects and then
 * run any attached shapes after them.
 */
abstract class PatternShape : Pattern() {
    override fun affectEntity(world: World, entity: Interactor, instance: Instance) {
        if (instance is ShapeInstance) {
            instance.effects.forEach { effect -> effect.run(world, entity) }
            instance.nextShape?.run(world, entity)
        }
    }

    override fun affectBlock(world: World, block: Interactor, instance: Instance) {
        if (instance is ShapeInstance) {
            instance.effects.forEach { effect -> effect.run(world, block) }
            instance.nextShape?.run(world, block)
        }
    }
}